sdk/test-utils/perf/src/perfTestBase.ts

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import minimist, { ParsedArgs as MinimistParsedArgs } from "minimist";
+import {
+  PerfOptionDictionary,
+  parsePerfOption,
+  DefaultPerfOptions,
+  defaultPerfOptions,
+  validateOptions,
+  ParsedPerfOptions
+} from "./options";
+import { PerfParallel } from "./parallel";
+import { AbortController } from "@azure/abort-controller";
+
+/**
+ * Defines the behavior of the PerfTest constructor, to use the class as a value.
+ */
+export interface PerfTestConstructor<
+  TOptions extends Record<string, unknown> = Record<string, unknown>
+> {
+  new (): PerfTestBase<TOptions>;
+}
+
+/**
+ * Conveys the structure of any Perf test.
+ * It allows developers to define the optional parameters to receive
+ * in a way that aids them to auto-complete these options while adding code inside of each method.
+ * It provides methods to parse the assigned options, as well as to set up and tear down any resources
+ * at a global level (only one setup and teardown for the whole execution of a Perf test),
+ * and at a local level, which happens once for each initialization of the test class
+ * (initializations are as many as the "parallel" command line parameter specifies).
+ */
+export abstract class PerfTestBase<TOptions = Record<string, unknown>> {
+  public abstract options: PerfOptionDictionary<TOptions>;
+
+  public get parsedOptions(): ParsedPerfOptions<TOptions & DefaultPerfOptions> {
+    // Only validate the options if they are defined: if (when) parsedOptions is called
+    // in the constructor, options will be undefined.
+    if (this.options) {
+      validateOptions({
+        ...this.options,
+        ...defaultPerfOptions
+      });
+    }
+
+    // This cast is needed because TS thinks
+    //   PerfOptionDictionary<TOptions & DefaultPerfOptions>
+    //   is different from
+    //   PerfOptionDictionary<TOptions> & PerfOptionDictionary<DefaultPerfOptions>
+    return parsePerfOption({
+      ...this.options,
+      ...defaultPerfOptions
```

Comment:
```suggestion
      ...defaultPerfOptions
      ...this.options,
```

Shouldn’t the default options go first? otherwise the provided options will be overwritten by the default ones 🤔

Reply:
they don't have overlap, order would not matter.

Reply:
wait, they don’t “have” to overlap means they “could” overlap. If they could overlap, order matters.

Reply:
if they can overlap, we could even write a test to ensure that the correct expected behavior is always satisfied.

Reply:
I wanted to write.. `they don't overlap` 😂

Reply:
Ideally, they should not overlap. 
In case they overlap, it's basically changing the behavior of the perf framework in unexpected ways.
I should probably add a check to make sure they do not overlap - either at the type level or at the validation level.

Will take this up in a different PR.
This snippet existed before as well, it has merely been moved from a different file.
I want to keep the changes scoped to the context of this PR.

Will handle the "no overlap" in a separate PR.

Reply:
Added in TODO